---
title: "Ribosomal Profiling"
author: "Mohanad Elchouemi"
date: "2/13/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Load the packages that you need. If they are not installed then don't forget to install them.

```{r}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# 
# BiocManager::install("ribor")
# install.packages("lmodel2")
# install.packages("devtools")
# install_github("ribosomeprofiling/ribor")

library("devtools")
library(ribor)
library(tidyr)
library(ggplot2)
library(FSA)
library(dplyr)
library(stringr)
library(janitor)
library(readxl)
library(ggplot2)
library("ggpubr")
library(lmodel2)
library(tidyverse)
```

Create the Ribo R object from the File I want to use.
```{r}
second.ribo <- Ribo("/Users/mohanadelchouemi/Desktop/Dr Cenik/RibosomeTranslation/Cenik_Lab_ME/all.ribo", rename = rename_default )
second.ribo 
```

### Used the GSM1691202 experiment because that is what corresponds to the Research article!

Getting the mRNA data
```{r}
rna <- get_rnaseq(second.ribo,
           tidy = TRUE,
           region = c( "CDS"),
           experiment = "GSM1691202",
           compact = FALSE,
           alias = TRUE) %>% mutate(GENE_NAME3 = str_replace(transcript , "-.*",replacement = "")) %>% 
  select(-c(transcript)) %>% select(count,GENE_NAME3) %>% rename(rna_count = count)

```

Getting the Ribosomal profiling footprint data
```{r}
rc <- get_region_counts(second.ribo,tidy = TRUE,
                        compact      = FALSE,
                        transcript  = FALSE,
                        alias       = TRUE,
                        region      = c("CDS"),
                        experiment ="GSM1691202") %>% 
  mutate(GENE_NAME3 = str_replace(transcript , "-.*",replacement = "")) %>% 
  select(-c(transcript)) %>% select(count,GENE_NAME3) %>%
  rename(rc_count = count)
```

Getting the proteomics data
```{r}
data <- read_xlsx(path = "/Users/mohanadelchouemi/Desktop/Dr Cenik/RibosomeTranslation/Cenik_Lab_ME/mmc4.xlsx", sheet = 2) %>% 
  mutate_at(vars(2:3),~as.numeric(.x ))
data[is.na(data)] <- 0
data$prot_count <- (data$`iBAQ 1 protein copies per cell` + data$`iBAQ 2 protein copies per cell`) / 2
prot <- data %>% select(`gene symbol`,prot_count) %>% rename(GENE_NAME3 = `gene symbol`)

transcripts_used_from_prot <- prot$GENE_NAME3

```
*transcripts_used* is a vector that contains all the transcripts/genes that are used in the experiment.

RNA-Seq vs Proteomics
```{r}
rna_prot <- left_join(x = rna,y = prot, by = "GENE_NAME3") %>% na.omit() %>% .[.$rna_count > 2 & .$prot_count >2,]
plot(log2(rna_prot$rna_count), log2(rna_prot$prot_count), xlab = "RNA-Seq", ylab = "Proteomics", main = "log2 CDS Read Counts", pch = 19, cex = 0.5)
abline(lm(log2(rna_prot$prot_count) ~ log2(rna_prot$rna_count), data = rna_prot), col = "blue")
cor(log2(rna_prot$rna_count), log2(rna_prot$prot_count), method = c("pearson"))
# Ranged Major Axis Regression for RNA-SEQ and Proteomics
lmodel2(log2(prot_count) ~ log2(rna_count),  data = rna_prot,range.y = "relative",range.x = "relative",nperm = 1000)
#the graph of this!
abline(-14.424822 ,3.5503772, col = "black")
```


Ribosomal Profiling Footprint(RPF) vs Proteomics
```{r}
rc_prot <- left_join(x = rc,y = prot, by = "GENE_NAME3") %>% na.omit() %>%
  .[.$rc_count > 2 & .$prot_count >2,]

plot(log2(rc_prot$rc_count), log2(rc_prot$prot_count), xlab = "Ribosomal Profiling FootPrint", ylab = "Proteomics", main = "log2 CDS Read Counts", pch = 19, cex = 0.5)

abline(lm(log2(rc_prot$prot_count) ~ log2(rc_prot$rc_count), data = rc_prot), col = "blue")

cor(log2(rc_prot$rc_count), log2(rc_prot$prot_count), method = c("pearson"))

# Ranged Major Axis Regression for RPF and Proteomics
lmodel2(log2(prot_count) ~ log2(rc_count),  data = rc_prot,range.y = "relative",range.x = "relative" )

#the graph of this!
abline(-5.471448,2.3189631 , col = "black")
```


# Setting Up the Environment prior to the FOR LOOP

# Taking into account the coverage data and acounting for outliers, could be due to errors.
```{r}
# A raw data frame of the ribosomal profiling data. Mainly used to get the right list of names.
rc_raw <- get_region_counts(second.ribo,tidy = TRUE,
                        compact      = FALSE,
                        transcript  = FALSE,
                        alias       = TRUE,
                        experiment ="GSM1691202") %>% 
  group_by(transcript) %>% 
  summarise(total_count = sum(count)) %>% 
  mutate(GENE_NAME3 = str_replace(transcript , "-.*",replacement = "")) %>% 
  filter(GENE_NAME3 %in% transcripts_used_from_prot)

?filter()

#19736 unique transcripts in the riboR file.
#2406 transcripts that are in proteomics and the ribosomal profiling data.

```
This narrows down the transcripts that will be used.

```{r}
rc_og <-get_region_counts(second.ribo,tidy = TRUE,
                        compact      = FALSE,
                        transcript  = FALSE,
                        alias       = TRUE,
                        experiment ="GSM1691202")
```
Generates the dataframe for the for loops to access the region counts information.


Pipeline that generates the coverage of the gene without any duplicates
```{r}
transcripts_names<- unique(rc_raw$transcript) # This is what will be iterated through for the for loop. 2,406 times
coord <- get_region_coordinates(second.ribo,alias = TRUE) %>% filter(transcript %in% transcripts_names) %>% select(transcript,CDS_start,CDS_stop)
```
coord is a dataframe that contains all the stop and start sites for the CDS.

The For Loop
```{r}
# Creating the tibble before starting the for loop.
cov_new <- tibble(
  transcript = NA,
  count = NA
)


#Pipeline that generates the coverage of the gene without 
# The loop in action
for (i in 1:length(transcripts_names)) {
  naming_variable = transcripts_names[i] 

  #cds_start<- coord[coord$transcript == naming_variable,]$CDS_start
  #cds_stop <- coord[coord$transcript == naming_variable,]$CDS_stop
  
  # Basically this function first creates a dataframe of the coverage of the gene. Then it makes the counts a numeric data type and deletes the old column that saved counts as character data type. Then it returns the CDS region that we get from the previous chunk using the region count coordinates. We then access that dataframe and get the start and stop CDS region. Then filter out outliers that are five or more standard devations away. Then we sum the CDS regions that are not an outlier then return that sum as a numeric and add it to a new tibble with transcript name and count.
  cov <- get_coverage(ribo.object = second.ribo,
                    name = naming_variable,
                    length      = TRUE,
                    alias       = TRUE,
                    tidy        = TRUE,
                    experiment = "GSM1691202",
                    compact = FALSE) %>% 
    mutate(position_num = as.numeric(position)) %>% select(-c(position)) %>%
  filter(position_num >= coord[coord$transcript == naming_variable,]$CDS_start & position_num <= coord[coord$transcript == naming_variable,]$CDS_stop) %>%
    filter(count > (mean(count) - 5*sd(count)) & count < (mean(count) + 5*sd(count))) %>%
    summarise(sum(count)) %>% as.numeric()
  
  cov_new <- add_row(.data = cov_new,transcript = naming_variable, count = cov)

  
}
# 
cov_new2 <-cov_new[-1,]
write.csv(cov_new2, file = "gene_coverage_perfect.csv")
```



 # This chunk is to address the issue of only using the CDS Region of each transcript before removing the outliers.
 
 #### https://stackoverflow.com/questions/8540143/add-consecutive-elements-of-a-vector-until-a-value
 lets see if this works lol.
 # https://bookdown.org/csgillespie/efficientR/programming.html
 
 ### Test case to see if the for loop works lol.
```{r}

cds_start<- coord[coord$transcript == "BAZ1A-203",]$CDS_start
cds_stop <- coord[coord$transcript == "BAZ1A-203",]$CDS_stop
cov <- get_coverage(ribo.object = second.ribo,
                    name = "BAZ1A-203",
                    length      = TRUE,
                    alias       = TRUE,
                    tidy        = TRUE,
                    experiment = "GSM1691202",
                    compact = FALSE) %>% mutate(position_num = as.numeric(position)) %>% select(-c(position)) %>%
  filter(position_num >= coord[coord$transcript == "BAZ1A-203",]$CDS_start & position_num <= coord[coord$transcript == "BAZ1A-203",]$CDS_stop) %>% 
  filter(count > (mean(count) - 5*sd(count)) & count < (mean(count) + 5*sd(count))) %>% summarise(value = sum(count)) %>% as.numeric()

#cov[1][1]

#mean(cov$count) + sd(cov$count)*5

#nom <- cov[cov$count > (mean(cov$count) - 5*sd(cov$count)) & cov$count < (mean(cov$count) + 5*sd(cov$count)),]

#cov_new <- add_row(.data = cov_new,transcript = transcript_names[i], count = nom)



# coord <- get_region_coordinates(second.ribo,alias = TRUE) %>% filter(transcript %in% transcripts_names) %>% select(transcript,CDS_start,CDS_stop)

```


Plotting out everything and seeing it all!
```{r}
rpf_edited <- read.csv("gene_coverage_perfect.csv")  %>%
  mutate(GENE_NAME3 = str_replace(transcript , "-.*",replacement = "")) %>%
  select(-c(transcript)) %>% select(count,GENE_NAME3) %>% rename(rpf_edited_count = count)

rpf_edited_prot <- left_join(x = rpf_edited,y = prot, by = "GENE_NAME3") %>% na.omit() %>% .[.$rpf_edited_count > 2 & .$prot_count >2,]

plot(log2(rpf_edited_prot$rpf_edited_count), log2(rpf_edited_prot$prot_count),
     xlab = "Ribosomal Profiling FootPrint edited", ylab = "Proteomics", main = "log2 Read Counts", pch = 19, cex = 0.5) +
abline(lm(log2(rpf_edited_prot$prot_count) ~ log2(rpf_edited_prot$rpf_edited_count), data = rpf_edited_prot), col = "blue") +
  abline(-7.092803,2.501529, col = "black")

cor(log2(rpf_edited_prot$rpf_edited_count), log2(rpf_edited_prot$prot_count), method = c("pearson"))

# Ranged Major Axis Regression for RPF and Proteomics
lmodel2(log2(prot_count) ~ log2(rpf_edited_count),  data = rpf_edited_prot,range.y = "relative",range.x = "relative",nperm = 1000)
#the graph of this!
#the abline is up above^^



```



Notes
# identify outliers,
# get residual plot
# function that gets (residuals)


Making a function that adds residuals to a dataframe and returns certain residuals away.
```{r}
# First make a model of your plots.
# This is gonna be from the adjusted RPF and Proteomics
library(modelr)  # Need modelr library to run this
model <- lm(log2(prot_count) ~ log2(rpf_edited_count), data = rpf_edited_prot)
plot(model)
rpf_edited_prot_residuals <- add_residuals(rpf_edited_prot, model, var = "resid") %>% filter(!between(resid,-5,5)) # final pipe returns the outliers!

# returns data frame with the outliers found! This makes log2 of the readcounts so can't rely on it!
residual_outlier_finder <- function(res_outliers,mydata,vector_y,vector_x){
 
  y <- mydata %>% pull(vector_y)
  x <- mydata %>% pull(vector_x)
  # creating a dataframe because the add_residual function needs some weird stuff.
  temp_df <- mydata %>% mutate(y = log2(y), x = log2(x))
  print(temp_df)
  model_func <- lm(y ~ x, data = temp_df)
  # returns a data frame that has all the genes that are 5 or more residuals away.
  return(temp_df %>% add_residuals(model_func,var = "resid") %>% filter(!between(resid,-res_outliers,res_outliers)) %>% select(-c(y,x)))
}

rpf_edited_prot %>% pull("prot_count")
boy <- residual_outlier_finder(res_outliers = 5, mydata = rpf_edited_prot,vector_y = "prot_count", vector_x = "rpf_edited_count")
?add_residuals
# index name and index
# extract column name



# Make a residual plot before and after
# Always check the accuracy of the function. 
# ggplot color them.
# check the ggplots.

model <- lm(log2(prot_count) ~ log2(rpf_edited_count), data = rpf_edited_prot)
plot(model)
rpf_edited_prot_residuals_test_case <- add_residuals(rpf_edited_prot, model, var = "resid") %>% mutate(label = ifelse(!between(resid,-5,5),"out","in")) %>% distinct() # final pipe returns the outliers!
# ggplot to see whether they fit or not.
rpf_edited_prot_residuals_test_case %>% ggplot(mapping = aes(log2(rpf_edited_count),log2(prot_count),color = label)) + geom_point() + geom_abline(intercept = 6.183,slope = 1.017) #+ geom_text(aes(label=ifelse(label == "out",as.character(GENE_NAME3),'')),hjust=0,vjust=0)
# Created a csv file without row number or quotations so that we can use the below website with no problem.

#http://llama.mshri.on.ca/funcassociate/
  
as.data.frame(boy$GENE_NAME3) %>% write.csv("for_web.csv",row.names = F,quote = F)

```


Gene Space File - All 2300 genes. 


What are the genes we are evaluating.

Mappability Open Python etc.
Read it in R.

This is a chunk to get mappabiltiy data frame and then graph this to the residuals.
```{r}
 test <- get_region_counts(second.ribo,tidy = TRUE,
                        compact      = FALSE,
                        transcript  = FALSE,
                        alias       = TRUE,
                        region      = c("CDS"),
                        length = FALSE,
                        experiment ="GSM1691202")
mappability <- read.csv("/Users/mohanadelchouemi/Desktop/Dr Cenik/RibosomeTranslation/Cenik_Lab_ME/pandaman.csv") %>% mutate(GENE_NAME3 = str_replace(X , "-.*",replacement = "")) %>% select(-X) %>% group_by(GENE_NAME3) %>% summarize(across(1:6,mean))

# Added some cases.... Figured out why because there are some names that have exact naming.
# mappability_residuals <- rpf_edited_prot_residuals_test_case %>% left_join(mappability) %>% count(GENE_NAME3) %>% arrange(-desc(n))

# Issues with "duplicates" because of removing the dash afterwards.
# Need to create a dataframe that takes into account what genes are those that are similar.

# Okay getting to work now.
mappability_residuals <- rpf_edited_prot_residuals_test_case %>% left_join(mappability) %>% pivot_longer(cols = contains("X"))

mappability_residuals %>% ggplot(mapping = aes(value,resid,color = label)) + geom_point() + facet_wrap(~name)


```
Preparing for the FuncAssociate
```{r}
#This is the list of genes that are the outliers!
as.data.frame(boy$GENE_NAME3) %>% write.csv("for_web.csv",row.names = F,quote = F)



rpf_edited_prot_residuals_test_case %>% 
  select(GENE_NAME3) %>% mutate(num = 1) %>% 
  write.table( file = "gene_space.tsv", row.names=FALSE, sep="\t",col.names = F, quote = F)



```


Okay so far: 
1) Ran Func Associate
2) Saw that "extracellular region part" are what the majority of the genes that are outliers on the Prot and RPF.
3) Going to create scatter plot of just Extra Cellular genes and everything else.
4) Mappability vs CDS length.

```{r}
extra_cellular_regions <- scan("omg.rtf",sep = " ",what = "character") %>% as.data.frame() %>% slice(10:811) %>% rename(E_R_P = ".") %>% pull(E_R_P) %>% as.data.frame() %>% slice(1:801) %>% pull(.) %>% append("ZMPSTE24",length(801)) %>% as.data.frame() %>% rename(genes = ".")
# Making a clean copy of the 802 genes that are form the extra cellular reigon parts
write.csv(x = extra_cellular_regions,file = "extra_cellular_region_parts_genes.csv")
#This will label the genes whether they are extra cellular or not.

rpf_edited_prot_distinction_between_extracellular_or_not <- rpf_edited_prot_residuals_test_case %>% mutate(ECRP = ifelse(GENE_NAME3 %in% extra_cellular_regions$genes,"yes","no")) 

# A plot that shows whether the data points are extra cellular region part or not, and whtether they are an outlier or not.
rpf_edited_prot_distinction_between_extracellular_or_not %>% select(c(1,2,3,5,6)) %>%
 ggplot(mapping = aes(log2(rpf_edited_count),log2(prot_count),color = label)) + geom_point() + facet_wrap(~ECRP)



# Now lets see if there is any difference in the correlation between teh two or not!
extracellular_vs_not <- rpf_edited_prot_distinction_between_extracellular_or_not %>% select(1,2,3,6)
extracellular <- extracellular_vs_not %>% filter(ECRP == "yes") %>% mutate(rpf_edited_count = rpf_edited_count %>% as.numeric())
not_extracellular <- extracellular_vs_not %>% filter(ECRP == "no") %>%  mutate(rpf_edited_count = rpf_edited_count %>% as.numeric())




# This is the correlation for extracellular region parts
plot(log2(extracellular$rpf_edited_count), log2(extracellular$prot_count), xlab = "Ribosomal Profiling FootPrint", ylab = "Proteomics", main = "log2 CDS Read Counts for ExtraCellular Region Parts", pch = 19, cex = 0.5) +
abline(lm(log2(extracellular$prot_count) ~ log2(extracellular$rpf_edited_count), data = extracellular), col = "blue") + abline(-5.404230,2.362228, col = "black")

cor(log2(extracellular$rpf_edited_count), log2(extracellular$prot_count), method = c("pearson"))

# Ranged Major Axis Regression for RPF and Proteomics
lmodel2(log2(prot_count) ~ log2(rpf_edited_count),  data = extracellular,range.y = "relative",range.x = "relative",nperm = 1000 )

#the graph of this!
# line of best fit of the RMA is above!




# This is the correlation for everything else

plot(log2(not_extracellular$rpf_edited_count), log2(not_extracellular$prot_count), xlab = "Ribosomal Profiling FootPrint", ylab = "Proteomics", main = "log2 CDS Read Counts for Not ExtraCellular Region Parts", pch = 19, cex = 0.5) +
abline(lm(log2(not_extracellular$prot_count) ~ log2(not_extracellular$rpf_edited_count), data = not_extracellular), col = "blue") + abline(-6.715800,2.4374465, col = "black")

cor(log2(not_extracellular$rpf_edited_count), log2(not_extracellular$prot_count), method = c("pearson"))

# Ranged Major Axis Regression for RPF and Proteomics
lmodel2(log2(prot_count) ~ log2(rpf_edited_count),  data = not_extracellular,range.y = "relative",range.x = "relative",nperm = 1000 )

#the graph of this!
# line of best fit of the RMA is above!



```

This is now showing mappability to CDS Length
```{r}

# Getting all the CDS for all the genes. Some genes are repreated since we removed the isoform label at the end. To combat this we found the mean of the two isoforms!
cds_lengths <- get_region_coordinates(second.ribo,alias = TRUE) %>% filter(transcript %in% transcripts_names) %>% select(transcript,CDS_start,CDS_stop) %>% 
  mutate(GENE_NAME3 = str_replace(transcript , "-.*",replacement = "")) %>% select(-1) %>% mutate(cds_lengths = CDS_stop - CDS_start) %>% select(3,4) %>% group_by(GENE_NAME3) %>% summarise(CDS_length = mean(cds_lengths))

cds_lengths_mappability <- left_join(cds_lengths,mappability)

cds_lengths_mappability %>% pivot_longer(cols = contains("X")) %>% ggplot(mapping = aes(log10(CDS_length),value)) + geom_point() + facet_wrap(~name)
```
 
CDS lengths vs residuals
2d density plots instead of geom point.
Ultimately we need to look at the sequences.


Verify that this is correct.

why are the residuals like that....
```{r}
resid_cds_length <- left_join(rpf_edited_prot_residuals_test_case, cds_lengths, by = "GENE_NAME3")

str(resid_cds_length)
ggplot(data = resid_cds_length, mapping = aes(x = log10(CDS_length),y= resid)) + geom_point()

ggplot(data = resid_cds_length, mapping = aes(x = log10(prot_count),y= resid)) + geom_point()

```
## Plotting Everything to see the patterns!
```{r}
# Get the lengths of the transcripts first
gene_lengths <- get_region_coordinates(second.ribo,alias = TRUE) %>% select(c(transcript,CDS_start,CDS_stop)) %>% mutate(CDS_length = CDS_stop - CDS_start) %>% mutate(GENE_NAME3 = str_replace(transcript , "-.*",replacement = "")) %>% select(c(GENE_NAME3,CDS_length)) %>% group_by(GENE_NAME3) %>% summarise(CDS_length = mean(CDS_length)) %>% ungroup()

# Get the ribosomal profiling data
ribo_profiling <- get_region_counts(second.ribo,tidy = TRUE,
                        compact      = FALSE,
                        transcript  = FALSE,
                        alias       = TRUE,
                        region      = c("CDS"),
                        experiment ="GSM1691202") %>% 
  mutate(GENE_NAME3 = str_replace(transcript , "-.*",replacement = "")) %>% 
  select(-c(transcript)) %>% select(count,GENE_NAME3) %>%
  rename(ribo_count = count)

# Get the proteomics data
data <- read_xlsx(path = "/Users/mohanadelchouemi/Desktop/Dr Cenik/RibosomeTranslation/Cenik_Lab_ME/mmc4.xlsx", sheet = 2) %>% 
  mutate_at(vars(2:3),~as.numeric(.x ))
data[is.na(data)] <- 0
data$prot_count <- (data$`iBAQ 1 protein copies per cell` + data$`iBAQ 2 protein copies per cell`) / 2
proteomics <- data %>% select(`gene symbol`,prot_count) %>% rename(GENE_NAME3 = `gene symbol`)

# Merge the two dataframes.  Proteomics on the left because we don't have that many proteomcis observations
proteomics_and_ribo_profiling <- left_join(proteomics,ribo_profiling, by = "GENE_NAME3") %>% na.omit()
proteomics_and_ribo_profiling <- proteomics_and_ribo_profiling[proteomics_and_ribo_profiling$ribo_count > 2 & proteomics_and_ribo_profiling$prot_count >2,]

# Now we can add the residuals. Residual = y - y hat ( observed minus expected)
# http://www.r-tutor.com/elementary-statistics/simple-linear-regression/residual-plot
library(modelr) # will use this package to add residuals easily
model2 <- lm(log2(prot_count) ~ log2(ribo_count), data = proteomics_and_ribo_profiling) # Need to create a model first
prot_ribo_resid <- add_residuals(proteomics_and_ribo_profiling, model2, var = "resid")

#Now we can combine prot_ribo_resid with the gene_lengths
prot_ribo_resid_lengths <- left_join(prot_ribo_resid, gene_lengths, by = "GENE_NAME3")

#Now we can plot everything and see what the relationships there are.


plot(log2(prot_ribo_resid_lengths$ribo_count), log2(prot_ribo_resid_lengths$prot_count), xlab = "ribo count", ylab = "prot count", main = "", pch = 19, cex = 0.5)



# ribo vs resid
plot(log2(prot_ribo_resid_lengths$ribo_count), model2$resid, xlab = "ribo count", ylab = "resid", main = "", pch = 19, cex = 0.5) 

plot(log2(prot_ribo_resid_lengths$prot_count), prot_ribo_resid_lengths$resid, xlab = "proteomics count", ylab = "resid", main = "", pch = 19, cex = 0.5) 

# These two are the same
plot(log2(prot_ribo_resid_lengths$CDS_length), prot_ribo_resid_lengths$resid, xlab = "CDS length count", ylab = "resid", main = "", pch = 19, cex = 0.5)

plot(log2(prot_ribo_resid_lengths$CDS_length), model2$resid, xlab = "CDS length count", ylab = "resid", main = "", pch = 19, cex = 0.5)

# Different model!
model3 <- lm(log2(prot_count)~log2(ribo_count) + log2(CDS_length) , data = prot_ribo_resid_lengths)

summary(model3)

plot(log2(proteomics_and_ribo_profiling$ribo_count), resid(model3), xlab = "ribo count", ylab = "resid", main = "lm(prot~ribo)", pch = 19, cex = 0.5) 

plot(log2(prot_ribo_resid_lengths$CDS_length),resid(model3))


plot(log2(prot_ribo_resid_lengths$CDS_length), log2(prot_ribo_resid_lengths$prot_count))
#################################################
#################################################
# Now I am gonna find the residuals using the RMA

lmodel2 <-lmodel2(log2(prot_count) ~ log2(ribo_count), data = proteomics_and_ribo_profiling,range.y = "relative",range.x = "relative",nperm = 1000)


# Pretty sure everything below here is wrong.

# The model formula is y = 2.3189631(X) - 5.471448	
# Going to calculate the residuals and also create my expected values based off of the RMA model!
prot_ribo_resids_RMA <- proteomics_and_ribo_profiling %>% mutate(expected_values_RMA = (2.3189631 *(log2(ribo_count)) - 5.471448)) %>% mutate(log2(ribo_count),log2(prot_count)) %>% mutate(residuals = log2(prot_count) - expected_values_RMA) %>% mutate(label = ifelse(!between(residuals,-5,5),"out","in")) %>% left_join(gene_lengths)

prot_ribo_resids_RMA %>%  ggplot(aes(x = log2(ribo_count),y = log2(prot_count),color = label)) +geom_abline(intercept = -5.471448, slope = 2.3189631) # geom_text(aes(label = GENE_NAME3))



```
Okay so now I have created my linear plot with the RMA and I will plug this data into funcAssociate! 
Didn't work ask why!!
```{r}
prot_ribo_resids_RMA %>% select(GENE_NAME3,label) %>% filter(label == "out") %>% select(GENE_NAME3) %>%pull %>%  length() #%>% write.csv("resids_out4_14_2021.csv",row.names = F,quote = F)
prot_ribo_resids_RMA %>%select(GENE_NAME3) %>% distinct() %>% mutate(num = 1) %>% 
  write.table( file = "gene_space_4_14_2021.tsv", row.names=FALSE, sep="\t",col.names = F, quote = F)
```

```{r}
ggplot(prot_ribo_resids_RMA,aes(log2(CDS_length),residuals,color = label)) + geom_point()

prot_ribo_resids_RMA %>% ggplot(aes(log2(ribo_count),residuals,color = label)) + geom_point()

prot_ribo_resids_RMA %>% filter(label == "in") %>% ggplot(aes(log2(ribo_count),residuals)) + geom_point()


ggplot(prot_ribo_resids_RMA,aes(log2(CDS_length),log2(ribo_count),color = label)) + geom_point() + facet_wrap(~label)#geom_hex(bins = 58)

ggplot(prot_ribo_resids_RMA,aes(log2(CDS_length),log2(prot_count),color = label)) + geom_point() + facet_wrap(~label)#geom_hex(bins = 58)



```
Create GLM to predict protein levels.
Linear model with two variables.
mulitple regession two continous variables.

lmodoel2 ......... one with error and the other with no error.... can i incorportate this with cds lengeth bc this has no erroer... its known for sure.

mappability vs residuals...
